cvmfs/export_plugin/fs_traversal.cc

```diff
@@ -0,0 +1,793 @@
+  ssize_t dest_iter = -1;
+  char * dest_entry = NULL;
+  struct cvmfs_attr *dest_st = cvmfs_attr_init();
+  dest->list_dir(dest->context_, dir, &dest_dir, &dest_len);
+  qsort(dest_dir, dest_len, sizeof(char *),  strcmp_list);
+
+  // While both have defined members to compare.
+  while (1) {
+    if (cmp <= 0) {
+      getNext(src, dir, src_dir, &src_entry, &src_iter,
+            &src_st, true, true, pstats);
+    }
+
+    if (cmp >= 0) {
+      getNext(dest, dir, dest_dir, &dest_entry, &dest_iter,
+              &dest_st, do_fsck, false, pstats);
+    } else {
+      // A destination entry was added
+      pstats->Lookup(SHRINKWRAP_STAT_DEST_ENTRIES)->Inc();
+    }
+
+    if (!src_entry && !dest_entry) {
+      // If we have visited all entries in both we break
+      break;
+    } else if (!src_entry) {
+      cmp = 1;
+    } else if (!dest_entry) {
+      cmp = -1;
+    } else {
+      cmp = strcmp(src_entry, dest_entry);
+    }
+
+    if (cmp <= 0) {
+      // Compares stats to see if they are equivalent
+      if (cmp == 0 && cvmfs_attr_cmp(src_st, dest_st, dest)
+        // Also check internal hardlink consistency in destination file system
+        // where applicable:
+        && (dest_st->cvm_checksum == NULL
+          || dest->is_hash_consistent(dest->context_, dest_st))) {
+        if (S_ISDIR(src_st->st_mode) && recursive) {
+          add_dir_for_sync(src_entry, recursive);
+        }
+        continue;
+      }
+      // If not equal, bring dest up-to-date
+      switch (src_st->st_mode & S_IFMT) {
+        case S_IFREG:
+          if (!handle_file(src, src_st, dest, dest_st, src_entry, pstats))
+            return false;
+          break;
+        case S_IFDIR:
+          if (!handle_dir(src, src_st, dest, dest_st, src_entry))
+            return false;
+          if (recursive)
+            add_dir_for_sync(src_entry, recursive);
+          break;
+        case S_IFLNK:
+          // Should likely copy the source of the symlink target
+          if (dest->do_symlink(dest->context_, src_entry,
+                               src_st->cvm_symlink, src_st) != 0) {
+            LogCvmfs(kLogCvmfs, kLogStderr,
+              "Traversal failed to symlink %s->%s : %d : %s",
+              src_entry, src_st->cvm_symlink, errno, strerror(errno));
+            return false;
+          }
+          break;
+        default:
+          LogCvmfs(kLogCvmfs, kLogStderr,
+            "Unknown file type for %s : %d", src_entry, src_st->st_mode);
+          return false;
+      }
+    /* Dest contains something missing from Src */
+    } else {
+      switch (dest_st->st_mode & S_IFMT) {
+        case S_IFREG:
+        case S_IFLNK:
+          if (dest->do_unlink(dest->context_, dest_entry) != 0) {
+            LogCvmfs(kLogCvmfs, kLogStderr,
+              "Failed to unlink file %s", dest_entry);
+            return false;
+          }
+          break;
+        case S_IFDIR:
+          Sync(dest_entry, src, dest, true, pstats, false);
```

Comment:
Add comment that this is recursive so that the contents of the directory are removed before trying to remove the directory. This could be handled other ways, but they are more complex. It may be worth the extra work to unlink in parallel if there is a large amount, but we haven't had a use case where this is true yet.